cache/models/samples.js

```diff
@@ -250,6 +267,142 @@ function applyFiltersOnSampObjs(sampObjArray, opts) {
   return filteredSamples;
 }
 
+/**
+ * Create properties array having and fields and values need to be set to
+ * update/create sample.
+ * @param  {Object} qbObj - Query body object
+ * @param  {Object} sampObj - Sample object
+ * @param  {Object} aspectObj - Aspect object
+ * @return {Array} Properties array need to be set
+ */
+function createSampHsetCommand(qbObj, sampObj, aspectObj) {
+  const props = [];
+  if (qbObj[sampFields.MSG_BODY]) {
+    props.push(sampFields.MSG_BODY, qbObj[sampFields.MSG_BODY]);
+  }
+
+  if (qbObj[sampFields.MSG_CODE]) {
+    props.push(sampFields.MSG_CODE, qbObj[sampFields.MSG_CODE]);
+  }
+
+  if (qbObj[sampFields.NAME]) {
+    props.push(sampFields.NAME, qbObj[sampFields.NAME]);
+  }
+
+  /* calculate status value, update status, previous status and changed at
+  fields in 2 cases: a) sampObj not present b) sample obj present and
+   previous status != current status. Always update status. */
+  let value;
+  if (qbObj[sampFields.VALUE]) {
+    value = qbObj[sampFields.VALUE];
+  } else if (!sampObj) {
+    value = '';
+  }
+
+  if (value) {
+    const status = sampleUtils.computeStatus(aspectObj, value);
+
+    if (!sampObj || (sampObj && (sampObj[sampFields.PRVS_STATUS] !== status))) {
+      const prevStatus = sampObj ? sampObj[sampFields.STATUS] :
+                            dbConstants.statuses.Invalid;
+      props.push(sampFields.PRVS_STATUS, prevStatus);
+      props.push(sampFields.STS_CHNGED_AT, new Date().toString());
+      props.push(sampFields.STATUS, status);
+    }
+  }
+
+  props.push(sampFields.UPD_AT, new Date().toString());
+  props.push(sampFields.VALUE, value);
+
+  // only in upsert, not bulk upsert,
+  let rlinks = [];
+  if (qbObj[sampFields.RLINKS]) {
+    rlinks = qbObj[sampFields.RLINKS];
+  } else {
+    rlinks = [];
+  }
+
+  props.push(sampFields.RLINKS, JSON.stringify(rlinks));
+  if (!sampObj) {
+    props.push(sampFields.CREATED_AT, new Date().toString());
+    props.push(sampFields.IS_DELETED, '0');
+
+    // swagger ensures that bulk upsert have names
+    props.push(sampFields.NAME, qbObj[sampFields.NAME]);
+  }
+
+  return props;
+}
+
+/**
+ * Upsert a sample. Check if subject exist. If yes, Get aspect and sample.
+ * If aspect exists, create fields array with values need to be set for sample
+ * in redis. Add aspect name to subject set, add aspect key to sample set and
+ * update/create sample hash. We use hset which updates a sample if exists,
+ * else creates a new one.
+ * @param  {Object} sampleQueryBodyObj - Query Body Object for a sample
+ * @returns {Array} - Array of redis commands responses
+ */
+function upsertOneSample(sampleQueryBodyObj) {
+  const sampleName = sampleQueryBodyObj.name;
+  const subjAspArr = sampleName.toLowerCase().split('|');
+  if (subjAspArr.length < TWO) {
+    throw new redisErrors.ResourceNotFoundError({
+      explanation: 'Incorrect sample name.',
```

Comment:
If we're throwing ResourceNotFoundError the explanation should be something like 'no sample was found with this name' but here it feels more like an "Invalid..." error because a sample name without a pipe character can never be valid, can never be "found".